include/arch/xtensa/cache.h

```diff
@@ -0,0 +1,70 @@
+/*
+ * Copyright 2021 Intel Corporation
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#ifndef ZEPHYR_INCLUDE_ARCH_XTENSA_CACHE_H_
+#define ZEPHYR_INCLUDE_ARCH_XTENSA_CACHE_H_
+
+#include <xtensa/config/core-isa.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#define Z_DCACHE_MAX (XCHAL_DCACHE_SIZE / XCHAL_DCACHE_WAYS)
+#define Z_DLINE (XCHAL_DCACHE_LINESIZE)
+
+#if XCHAL_DCACHE_SIZE
+#define Z_IS_POW2(x) (((x) > 0) && (((x) & ((x)-1)) == 0))
+BUILD_ASSERT(Z_IS_POW2(Z_DLINE));
+BUILD_ASSERT(Z_IS_POW2(Z_DCACHE_MAX));
+#endif
+
+static inline void z_xtensa_cache_flush(void *addr, size_t bytes)
+{
+#if XCHAL_DCACHE_SIZE
+	char *start = (char *)addr;
+
+	for (char *end = start + bytes; start <= end; start += Z_DLINE) {
```

Comment:
Yeah, the idea here was to favor code size and performance over strict adherence to flush boundaries and assume that the user architecture didn't rely on flush timing.  But I can see arguments either way; note that invalidate does it right because that process causes bugs if it's not exact.